internal/endpoint/pop3/pop3.go

```diff
@@ -0,0 +1,466 @@
+	}
+
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(false, nil, []imap.FetchItem{imap.FetchRFC822Size}, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return 0, 0, err
+	}
+
+	count := 0
+	size := 0
+	for msg := range(c) {
+		count += 1
+		size += int(msg.Size)
+	}
+
+	return count, size, nil
+}
+
+// List of sizes of all messages in bytes (octets)
+func (endp *Endpoint) List(user pop3backend.User) (octets []int, err error) {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return nil, err
+	}
+
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(false, nil, []imap.FetchItem{imap.FetchRFC822Size}, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return nil, err
+	}
+
+	items := make([]int, 0)
+	for msg := range(c) {
+		items = append(items, int(msg.Size))
+	}
+
+	return items, nil
+}
+
+// Returns whether message exists and if yes, then return size of the message in bytes (octets)
+func (endp *Endpoint) ListMessage(user pop3backend.User, msgId int) (exists bool, octets int, err error) {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return false, 0, err
+	}
+
+	seqset := imap.SeqSet{}
+	seqset.AddNum(uint32(msgId))
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(true, &seqset, []imap.FetchItem{imap.FetchRFC822Size}, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return false, 0, err
+	}
+
+	msg, ok := <- c
+	if !ok {
+		return false, 0, nil
+	}
+
+	return true, int(msg.Size), nil
+}
+
+// Retrieve whole message by ID - note that message ID is a message position returned
+// by List() function, so be sure to keep that order unchanged while client is connected
+// See Lock() function for more details
+func (endp *Endpoint) Retr(user pop3backend.User, msgId int) (message string, err error) {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return "", err
+	}
+
+	seqset := imap.SeqSet{}
+	seqset.AddNum(uint32(msgId))
+	c := make(chan *imap.Message)
+	err = mailbox.ListMessages(true, &seqset, []imap.FetchItem{imap.FetchRFC822Size}, c)
+	if err != nil && err != mess.ErrNoMessages {
+		return "", err
+	}
+
+	msg, ok := <- c
+	if !ok {
+		return "", fmt.Errorf("internal server error")
+	}
+
+	return strconv.FormatUint(uint64(msg.Uid), 10), nil
+}
+
+// Delete message by message ID - message should be just marked as deleted until
+// Update() is called. Be aware that after Dele() is called, functions like List() etc.
+// should ignore all these messages even if Update() hasn't been called yet
+func (endp *Endpoint) Dele(user pop3backend.User, msgId int) error {
+	mailbox, err := endp.getMailbox(user)
+	if err != nil {
+		return err
+	}
+
+	seqset := imap.SeqSet{}
+	seqset.AddNum(uint32(msgId))
+	return mailbox.UpdateMessagesFlags(true, &seqset, imap.SetFlags, false, []string{imap.DeletedFlag})
+}
+
+// Undelete all messages marked as deleted in single connection
+func (endp *Endpoint) Rset(user pop3backend.User) error {
+	return fmt.Errorf("pop3: unimplemented")
```

Comment:
Not sure how this can be implemented. Perhaps keep track of all messages that were Deleted and revert if this is called?

Reply:
That seems to be most reliable solution, considering current IMAP-based interface.